scipy/optimize/zeros.py

```diff
@@ -209,6 +211,93 @@ def newton(func, x0, fprime=None, args=(), tol=1.48e-8, maxiter=50,
     raise RuntimeError(msg)
 
 
+def _array_newton(func, x0, fprime, args, tol, maxiter, fprime2,
+                  failure_idx_flag=False):
+    """
+    A vectorized version of Newton, Halley, and secant methods for arrays. Do
+    not use this method directly. This method is called from :func:`newton`
+    when ``np.isscalar(x0)`` is true. For docstring, see :func:`newton`.
+    """
+    p0 = np.asarray(x0)  # convert to ndarray
+    failures = np.ones(p0.shape, dtype=bool)  # at start, nothing converged
+    if fprime is not None:
+        # Newton-Raphson method
+        for iteration in range(maxiter):
+            myargs = (p0,) + args
+            fder = np.asarray(fprime(*myargs))  # convert to ndarray
+            zero_der = (fder == 0)
+            if zero_der.all():
+                msg = "all derivatives were zero."
+                warnings.warn(msg, RuntimeWarning)
+                if failure_idx_flag:
+                    p0 = (p0, failures, zero_der)
+                return p0
+            fval = np.asarray(func(*myargs))  # convert to ndarray
+            newton_step = fval / fder
+            fder2 = 0 if fprime2 is None else np.asarray(fprime2(*myargs))
+            # if fder is zero, warns "RuntimeWarning: divide by zero"
+            p = p0 - newton_step / (1.0 - 0.5 * newton_step * fder2 / fder)
+            p = np.where(zero_der, p0, p)  # where zero der, use last guess
+            failures = np.abs(p - p0) >= tol  # items not yet converged
+            # stop iterating if there aren't any failures, not incl zero der
+            if not failures[~zero_der].any():
```

Comment:
This return code is similar enough to the one a few lines before to consider having a single version of them. Why not break out of the loop and have the code just once after all the iterations?